### PR TITLE
Add `OAuth2ErrorResponseErrorHandler` for `restTemplate` used in AzureAD

### PR DESCRIPTION
**Pull request by chenrujun**
1. Add `OAuth2ErrorResponseErrorHandler` for `restTemplate` used in `AzureAD`.

2. Use 2 `restTemplate` bean. Because `OAuth2ErrorResponseErrorHandler` is required when retrieve an access token. When retrieve key to validate access token. Then `OAuth2ErrorResponseErrorHandler` must not be used.

### CODE REVIEW

**Comment by chenrujun** [comment]
Hi, @saragluna , @moarychan , please help to review this PR when you have time.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aadb2c/configuration/AadB2cOAuth2ClientConfiguration.java [hunk 1]
```diff
@@ -131,48 +134,51 @@ private ClientRegistration buildClientRegistration(Map.Entry<String, Authorizati
...
+                .provider(azurePasswordProvider(restOperations))
                 .build();
         DefaultOAuth2AuthorizedClientManager manager = new DefaultOAuth2AuthorizedClientManager(clients, authorizedClients);
         manager.setAuthorizedClientProvider(authorizedClientProvider);
         return manager;
     }
 
-    private RefreshTokenOAuth2AuthorizedClientProvider azureRefreshTokenProvider(RestOperations aadAuthRestOperations) {
+    private RefreshTokenOAuth2AuthorizedClientProvider azureRefreshTokenProvider(
+            @Qualifier(AZURE_AD_ACCESS_TOKEN_RETRIEVER_REST_OPERATIONS_BEAN_NAME) RestOperations restOperations) {
```
Comment (moarychan) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aadb2c/configuration/AadB2cOAuth2ClientConfiguration.java:160]: No @Qualifier annotation is required.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aadb2c/configuration/AadB2cOAuth2ClientConfiguration.java [hunk 2]
```diff
@@ -131,48 +134,51 @@ private ClientRegistration buildClientRegistration(Map.Entry<String, Authorizati
...
         DefaultRefreshTokenTokenResponseClient responseClient = new DefaultRefreshTokenTokenResponseClient();
-        responseClient.setRestOperations(aadAuthRestOperations);
+        responseClient.setRestOperations(restOperations);
         provider.setAccessTokenResponseClient(responseClient);
         return provider;
     }
 
-    private ClientCredentialsOAuth2AuthorizedClientProvider azureClientCredentialProvider(RestOperations aadAuthRestOperations) {
+    private ClientCredentialsOAuth2AuthorizedClientProvider azureClientCredentialProvider(
+            @Qualifier(AZURE_AD_ACCESS_TOKEN_RETRIEVER_REST_OPERATIONS_BEAN_NAME) RestOperations restOperations) {
```
Comment (moarychan) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aadb2c/configuration/AadB2cOAuth2ClientConfiguration.java:169]: the same.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AadOAuth2ClientConfiguration.java [hunk 3]
```diff
@@ -160,31 +163,31 @@ RefreshTokenOAuth2AuthorizedClientProvider azureRefreshTokenProvider(
...
+            client.setRestOperations(restOperations);
             client.setRequestEntityConverter(converter);
 
             builder.accessTokenResponseClient(client);
         }
     }
 
     private void clientCredentialsGrantBuilderAccessTokenResponseClientCustomizer(
-            RestOperations aadAuthRestOperations,
+            @Qualifier(AZURE_AD_ACCESS_TOKEN_RETRIEVER_REST_OPERATIONS_BEAN_NAME) RestOperations restOperations,
```
Comment (moarychan) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AadOAuth2ClientConfiguration.java:190]: No @qualifier annotation is required.

**Review by moarychan — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java [hunk 4]
```diff
@@ -29,14 +31,62 @@
...
+     *
+     * @param builder The RestTemplateBuilder bean, which can be provided in {@link RestTemplateAutoConfiguration}
+     *                if customer not provide this bean.
+     * @return RestOperations bean
+     * @see AadOAuth2ClientConfiguration
+     * @see AadB2cOAuth2ClientConfiguration
+     */
+    @Bean(AZURE_AD_ACCESS_TOKEN_RETRIEVER_REST_OPERATIONS_BEAN_NAME)
+    @ConditionalOnMissingBean(name = AZURE_AD_ACCESS_TOKEN_RETRIEVER_REST_OPERATIONS_BEAN_NAME)
+    public RestOperations azureAdAccessTokenRetrieverRestOperations(RestTemplateBuilder builder) {
```
[the remark this answers is not in the document]
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java:74]: In scenarios like `resource_server_with_obo` and `web_application_and_resource_server`, both the 2 bean should be created.

image

[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java:74]: > In scenarios like `resource_server_with_obo` and `web_application_and_resource_server`, both the 2 bean should be created.

Yes, the rest operation bean can be used for `web_application` or `resource_server` respectively, I'm not sure it's necessary to create on demand.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AadOAuth2ClientConfiguration.java [hunk 5]
```diff
@@ -101,7 +104,7 @@ OAuth2ClientAuthenticationJwkResolver oAuth2ClientAuthenticationJwkResolver(AadA
     @Bean
     @ConditionalOnMissingBean
     OAuth2AuthorizedClientManager authorizedClientManager(
-            RestOperations aadAuthRestOperations,
+            @Qualifier(AZURE_AD_ACCESS_TOKEN_RETRIEVER_REST_OPERATIONS_BEAN_NAME) RestOperations restOperations,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AadOAuth2ClientConfiguration.java:107]: Inject this bean through the constructor. We don't need to add qualifier everywhere in this class.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AadResourceServerConfiguration.java [hunk 6]
```diff
@@ -49,18 +52,18 @@ public class AadResourceServerConfiguration {
...
-     * @param aadAuthRestOperations the RestOperation used to connect Azure AD
-     * @return Get the jwtDecoder instance.
+     * @param restOperations the restOperations used to get key to validate access token
+     * @return Get the jwtDecoder instance
      */
     @Bean
     @ConditionalOnMissingBean(JwtDecoder.class)
     public JwtDecoder jwtDecoder(AadAuthenticationProperties aadAuthenticationProperties,
-                                 RestOperations aadAuthRestOperations) {
+                                 @Qualifier(AZURE_AD_ACCESS_TOKEN_VALIDATOR_REST_OPERATIONS_BEAN_NAME) RestOperations restOperations) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AadResourceServerConfiguration.java:61]: Same here, I suggest we inject this bean via ctor.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java [hunk 7]
```diff
@@ -29,51 +33,77 @@
 public class AadRestOperationConfiguration {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java:33]: I suggest we split this class into two classes, `AadOAuth2ClientRestOperationConfiguration` and `AadResourceServerRestOperationConfiguration`.

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java:33]: And on `AadOAuth2ClientConfiguration` import this first one, on `AadResourceServerConfiguration` import the other. I agree with what @moarychan said, when only either oauth2 client or resource server is used, we don't need both beans in the context.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java [hunk 8]
```diff
@@ -29,51 +33,77 @@
...
+            "azureAdAccessTokenRetrieverRestOperations";
+
+    /**
+     * Bean name of RestOperation bean. The bean is used to get key to validate access token.
+     *
+     * @see AadRestOperationConfiguration#azureAdAccessTokenValidatorRestOperations(RestTemplateBuilder)
+     */
+    public static final String AZURE_AD_ACCESS_TOKEN_VALIDATOR_REST_OPERATIONS_BEAN_NAME =
+            "azureAdAccessTokenValidatorRestOperations";
+
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java:50]: Better put these bean names in `AadResourceServerConfiguration`, and we better call them by wider scenario, like `oauth2 client` and `resource server`

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aadb2c/AadB2cResourceServerAutoConfiguration.java [hunk 9]
```diff
@@ -74,27 +77,28 @@ public AadTrustedIssuerRepository trustedIssuerRepository() {
...
+     * @param restOperations the rest operations used to get key to validate access token
      * @param aadTrustedIssuerRepository the AAD trusted issuer repository
      * @param resourceRetriever the resource retriever
      * @return JWTClaimsSetAwareJWSKeySelector bean
      */
     @Bean
     @ConditionalOnMissingBean
     public JWTClaimsSetAwareJWSKeySelector<SecurityContext> aadIssuerJwsKeySelector(
-        RestOperations restOperations,
+            @Qualifier(AZURE_AD_ACCESS_TOKEN_VALIDATOR_REST_OPERATIONS_BEAN_NAME) RestOperations restOperations,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aadb2c/AadB2cResourceServerAutoConfiguration.java:101]: Inject this bean via ctor.

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadRestOperationConfiguration.java:74]: It won't be a heavy cost to create a bean even if it's never used.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2ClientRestOperationConfigurationTest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import static com.azure.spring.cloud.autoconfigure.aad.implementation.AadOauth2RestOperationConfigurationTestUtil.hasMessageConvertersForAccessToken;
+import static com.azure.spring.cloud.autoconfigure.aad.implementation.AadOauth2RestOperationConfigurationTestUtil.hasOAuth2ErrorResponseErrorHandler;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class AadOauth2ClientRestOperationConfigurationTest {
+
+    @Test
+    public void testRestOperationBeanConfiguration() {
+        new ApplicationContextRunner()
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2ClientRestOperationConfigurationTest.java:21]: We can remove these two `public`, test classes and methods can be package-private.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2ResourceServerRestOperationConfigurationTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+import static com.azure.spring.cloud.autoconfigure.aad.implementation.AadOauth2ResourceServerRestOperationConfiguration.AAD_OAUTH_2_RESOURCE_SERVER_REST_OPERATION_BEAN_NAME;
+import static com.azure.spring.cloud.autoconfigure.aad.implementation.AadOauth2RestOperationConfigurationTestUtil.hasMessageConvertersForAccessToken;
+import static com.azure.spring.cloud.autoconfigure.aad.implementation.AadOauth2RestOperationConfigurationTestUtil.hasOAuth2ErrorResponseErrorHandler;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class AadOauth2ResourceServerRestOperationConfigurationTest {
+
+    @Test
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2ResourceServerRestOperationConfigurationTest.java:19]: Same here.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2RestOperationConfigurationTestUtil.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.springframework.http.converter.FormHttpMessageConverter;
+import org.springframework.http.converter.HttpMessageConverter;
+import org.springframework.security.oauth2.client.http.OAuth2ErrorResponseErrorHandler;
+import org.springframework.security.oauth2.core.http.converter.OAuth2AccessTokenResponseHttpMessageConverter;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.List;
+
+public class AadOauth2RestOperationConfigurationTestUtil {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2RestOperationConfigurationTestUtil.java:14]: Same

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2RestOperationConfigurationTestUtil.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+
+    static boolean hasOAuth2ErrorResponseErrorHandler(RestTemplate restTemplate) {
+        return restTemplate.getErrorHandler() instanceof OAuth2ErrorResponseErrorHandler;
+    }
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2RestOperationConfigurationTestUtil.java:19]: We don't need this static method. Inlining this in your method will make it more readable.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2RestOperationConfigurationTestUtil.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    static boolean hasFormHttpMessageConverter(List<HttpMessageConverter<?>> converters) {
+        return converters.stream()
+                .anyMatch(converter -> converter instanceof FormHttpMessageConverter);
+    }
+
+    static boolean hasOAuth2AccessTokenResponseHttpMessageConverter(List<HttpMessageConverter<?>> converters) {
+        return converters.stream()
+                .anyMatch(converter -> converter instanceof OAuth2AccessTokenResponseHttpMessageConverter);
+    }
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/AadOauth2RestOperationConfigurationTestUtil.java:34]: If this is a util class, we can try to make this method more generic, 

```java
static <T> boolean hasXXX(List<T> list, Class<> clazz) {
        return list.stream()
                .anyMatch(converter -> xxx);
    }

```

**Review by saragluna — CHANGES_REQUESTED** [review]

**Comment by chenrujun** [comment]
FYI: PR to update reference doc: https://github.com/MicrosoftDocs/azure-dev-docs-pr/pull/3570

**Review by saragluna — APPROVED** [review]

**Pull request merged**